2016/proposals/rai-vivek.md

```diff
@@ -0,0 +1,360 @@
+
+An additional objective of this section is to keep the design sweet and simple
+(KISS).
+
+### Collectable information
+
+In this section, we discuss the parameters that can be collected by the SwC’s
+installation-test scripts to help us analyze and improve installation setup for
+future workshops.
+
+The goal is to collect anonymized, non-identifiable machine information and
+status report of the installation-test scripts.
+
+#### Machine information
+
+* Platform (or System)
+* Architecture (or Machine)
+* Node (Hostname), Processor
+* OS (or Distribution), Release version, Name
+
+Most of this information can be collected using the `platform` Python module.
+A snippet of useful commands is shown below.
+
+```
+import platform
+platform.uname() # cross platform
+platform.win32_ver() # windows
+platform.mac_ver() # mac
+platform.linux_distribution() # linux
+```
+
+#### Status report
+
+* Software check status (pass/fail) along with relevant messages
+* Version number of the found (installed) softwares
+* Any error that may have occurred in the script itself
+* Process information
+* Timestamp
+* (Optional) User email
+
+
+The current installation-test script implementation already collects the list of
+*successful* and *failed* checks along with version information and error
+message as available.
+
+### Server
+
+The server will collect the diagnostic report submitted by installation-scripts
+when they are run on the user machine.
+
+A simple design of the setup is discussed below:
+
+#### API
+
+The following steps are involved in the process:
+
+* Installation-script sends a `POST` request to the server endpoint (for example, https://installation.software-carpentry.org/report).
+  > Using Flask web-development framework for Python. A robust, flexible
+  > choice that is easy to develop and maintain. Scalability is not an issue for
+  > our use case.
+
+* Server verifies the data, open database connection, updates the record.
+  > As suggested in the ideas sections, SQLite will be a low-barrier, easier to
+  > maintain database, appropriate for our needs.
+
+* Send a status response back to the client. The response can contain feedback,
+  suggestions or perhaps link to some survey.
+
+* Ability to export diagnostic data into JSON or CSV format.
+  > It’ll allow us to separate the user interface from backend creating a modular
+  > setup.
+
+#### Database
+
+Based on the current nature of collectable information, write-volume and
+functionality required SQLite seems to be a fit choice. In this regard, the
+database schema is an important component that determines how we are going to
+store the collected data in the database.
+
+The following illustrates a typical schema that might be used:
+
+```
+TABLE machine_info {
+  id PRIMARY KEY AUTO,
+  system VARCHAR,
+  hostname VARCHAR,
+  release VARCHAR,
+  version VARCHAR,
+  machine VARCHAR,
+  processor VARCHAR,
+  email VARCHAR,
+  time VARCHAR,
```

Comment:
Okay; you could write the schema using declarative_base from SQLAlchemy (ie. model as a class).